File: backend/app/schemas/product.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.product import ProductCategory


class ProductBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    category: ProductCategory
    price: float = Field(..., gt=0)
    quantity: float = Field(..., ge=0)
    unit: str = Field(default="kg", max_length=32)
    image_url: str | None = None
    image_urls: list[str] | None = None  # Multiple images


class ProductCreate(ProductBase):
    pass


class ProductUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    category: ProductCategory | None = None
    price: float | None = Field(None, gt=0)
    quantity: float | None = Field(None, ge=0)
    unit: str | None = Field(None, max_length=32)
    image_url: str | None = None
    is_active: bool | None = None


class ProductResponse(ProductBase):
    id: UUID
    farmer_id: UUID
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
    
    def model_dump(self, *, mode='python', **kwargs):
        """Override to ensure enum values are serialized as strings for JSON"""
        data = super().model_dump(mode=mode, **kwargs)
        # Convert enum to its value for JSON serialization
        if mode == 'json' or 'json' in str(kwargs.get('mode', '')):
            if 'category' in data and hasattr(data['category'], 'value'):
                data['category'] = data['category'].value
        return data


class ProductListResponse(BaseModel):
    items: list[ProductResponse]
    total: int

File: mobile/lib/features/products/models/product.dart
import 'package:equatable/equatable.dart';

class Product extends Equatable {
  const Product({
    required this.id,
    required this.farmerId,
    required this.name,
    required this.description,
    required this.category,
    required this.price,
    required this.quantity,
    required this.unit,
    required this.isActive,
    required this.imageUrl,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as String,
      farmerId: json['farmer_id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      category: json['category'] as String,
      price: (json['price'] as num).toDouble(),
      quantity: (json['quantity'] as num).toDouble(),
      unit: json['unit'] as String? ?? 'kg',
      isActive: json['is_active'] as bool? ?? true,
      imageUrl: json['image_url'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'farmer_id': farmerId,
      'name': name,
      'description': description,
      'category': category,
      'price': price,
      'quantity': quantity,
      'unit': unit,
      'is_active': isActive,
      'image_url': imageUrl,
    };
  }

  final String id;
  final String farmerId;
  final String name;
  final String? description;
  final String category;
  final double price;
  final double quantity;
  final String unit;
  final bool isActive;
  final String? imageUrl;
  final DateTime createdAt;
  final DateTime updatedAt;

  @override
  List<Object?> get props => [id, name, category, price, quantity, unit, imageUrl, isActive, farmerId, createdAt, updatedAt];
}
